src/server/hueevents.rs

```diff
@@ -0,0 +1,83 @@
+use std::collections::VecDeque;
+
+use chrono::{DateTime, Utc};
+use tokio::sync::broadcast::Sender;
+
+use crate::hue::event::EventBlock;
+
+#[derive(Clone, Debug)]
+pub struct HueEventRecord {
+    timestamp: DateTime<Utc>,
+    index: u32,
+    pub block: EventBlock,
+}
+impl HueEventRecord {
+    pub fn id(&self) -> String {
+        format!("{}:{}", self.timestamp.timestamp(), self.index)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct HueEventStream {
+    prev_ts: DateTime<Utc>,
+    prev_index: u32,
+    hue_updates: Sender<HueEventRecord>,
+    buffer: VecDeque<HueEventRecord>,
+}
+
+impl HueEventStream {
+    pub fn new(buffer_capacity: usize) -> Self {
+        Self {
+            prev_ts: Utc::now(),
+            prev_index: 0,
+            hue_updates: Sender::new(32),
+            buffer: VecDeque::with_capacity(buffer_capacity),
+        }
+    }
+
+    fn add_to_buffer(&mut self, record: HueEventRecord) {
+        if self.buffer.len() == self.buffer.capacity() {
+            self.buffer.pop_front();
+            self.buffer.push_back(record);
+            debug_assert!(self.buffer.len() == self.buffer.capacity());
+        } else {
+            self.buffer.push_back(record);
+        }
+    }
+
+    fn generate_record(&mut self, block: EventBlock) -> HueEventRecord {
+        let ts = Utc::now();
+        if ts.timestamp() == self.prev_ts.timestamp() {
+            self.prev_index += 1;
+        } else {
+            self.prev_index = 0;
+            self.prev_ts = ts;
+        }
+        HueEventRecord {
+            block,
+            timestamp: ts,
+            index: self.prev_index,
+        }
```

Comment:
I really like this new helper struct :)

After this refactoring, I think `prev_index` doesn't make so much sense anymore. We should rename to just `index`.

If we also rename `ts` -> `timestamp`, then the last line will be just

```rust
HueEventRecord { block, timestamp, index }
```

(maybe also `prev_ts` -> `timestamp`?)